auth/oidc.go

```diff
@@ -439,3 +533,180 @@ func SetURLQueryParam(strURL, name, value string) (string, error) {
+	metadataChecksum := hex.EncodeToString(hash.Sum(nil))
+
+	if op.metadataChecksum == metadataChecksum {
+		base.Debugf(base.KeyAuth, "No change in discovery config detected at this time, next sync will be at %v", expiresAt)
+		return op.nextSyncAfter(expiresAt), nil
+	}
+	op.metadataChecksum = metadataChecksum
+
+	var metadata ProviderMetadata
+	if err := base.JSONUnmarshal(bodyBytes, &metadata); err != nil {
+		base.Debugf(base.KeyAuth, "Error parsing body during discovery sync: %v", err)
+		return 0, err
+	}
+
+	if metadata.Issuer != op.Issuer {
+		return 0, fmt.Errorf("oidc: issuer did not match the issuer returned by provider, expected %q got %q", op.Issuer, metadata.Issuer)
+	}
+	verifier := op.GenerateVerifier(&metadata, context.Background())
+	op.client.SetVerifier(verifier)
+	op.client.SetEndpoint(metadata.Endpoint())
+	return op.nextSyncAfter(expiresAt), nil
+}
+
+// nextSyncAfter returns the elapsed time between now and the next provider
+// metadata sync time. If the metadata expiry is zero or grater than 24 hours, the
+// next sync will occur in 24 hours. If the expiry is grater less than 1 minute,
+// the next sync will occur in next minute. If expiry falls in between 1 minute and
+// 24 hours, the next sync will occur at that interval from now.
+func (op *OIDCProvider) nextSyncAfter(exp time.Time) time.Duration {
+	if exp.IsZero() {
+		return MaximumProviderConfigSyncInterval
+	}
+	t := exp.Sub(time.Now())
+	if t > MaximumProviderConfigSyncInterval {
+		t = MaximumProviderConfigSyncInterval
+	} else if t < MinimumProviderConfigSyncInterval {
+		t = MinimumProviderConfigSyncInterval
+	}
+	return t
+}
+
+// StopDiscoverySync stops the provider metadata discovery sync
+// if it has started against this provider.
+func (op *OIDCProvider) StopDiscoverySync() {
+	if op.stopDiscoverySync != nil {
+		close(op.stopDiscoverySync)
+	}
+}
+
+// Returns the value of max-age directive from the Cache-Control HTTP header, i.e. the maximum
+// amount of time in seconds that fetched responses are allowed to be used again (from the time
+// when a request is made). The second value (ok) is true if a non max-age exists, and false if not.
+func cacheControlMaxAge(header http.Header) (maxAge time.Duration, ok bool, err error) {
+	for _, field := range strings.Split(header.Get("Cache-Control"), ",") {
+		parts := strings.SplitN(strings.TrimSpace(field), "=", 2)
+		k := strings.ToLower(strings.TrimSpace(parts[0]))
+		if k != "max-age" {
+			continue
+		}
+		if len(parts) == 1 {
+			return 0, false, errors.New("max-age has no value")
+		}
+		v := strings.TrimSpace(parts[1])
+		if v == "" {
+			return 0, false, errors.New("max-age has empty value")
+		}
+		age, err := strconv.Atoi(v)
+		if err != nil {
+			return 0, false, err
+		}
+		if age <= 0 {
+			return 0, false, nil
+		}
+		return time.Duration(age) * time.Second, true, nil
+	}
+	return 0, false, nil
+}
+
+// getExpiry returns time-to-live, the number of nano seconds after which the cached
```

Comment:
This comment doesn't indicate that this is using the Expires header - that seems relevant.  Information on why we need to convert (Date + Expires) headers to ttl would be useful as well.